File: flink/flink_kafka_java_demo_1/src/main/java/demo/flink/function/StringMessageProcessFunction.java
package demo.flink.function;

import com.typesafe.config.Config;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;

public class StringMessageProcessFunction extends ProcessFunction<String, String> implements Serializable {

    private Config jobConfig;

    private Logger logger = LoggerFactory.getLogger(StringMessageProcessFunction.class);

    public StringMessageProcessFunction(Config jobConfig){
        this.jobConfig = jobConfig;
    }
    @Override
    public void processElement(String s, ProcessFunction<String, String>.Context context, Collector<String> collector) throws Exception {
        System.out.println("StringMessageProcessFunction :: processElement() :: start");
        logger.info("Message Received From Kafka: " + s);
        System.out.println("Message Received From Kafka: " + s);
        System.out.println("StringMessageProcessFunction :: processElement() :: end");
    }
}

File: flink/flink_kafka_scala_demo_1/src/main/scala/demo/flink/scala/one/function/StringMessageProcessFunction.scala
package demo.flink.scala.one.function


import demo.flink.scala.one.config.AppConfig
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.util.Collector

class StringMessageProcessFunction(config: AppConfig)(implicit val stringTypeInfo: TypeInformation[String]) extends ProcessFunction[String, String]{
  override def processElement(i: String, context: ProcessFunction[String, String]#Context, collector: Collector[String]): Unit = {
    println("kafka message received : "+ i)
    context.output(config.messageTransmitterTag, i.toUpperCase)
    println("Message forwarded to output topic after processing.")
  }
}
